### ISSUE TITLE
Implement elm-fates fire design doc to bring to parity with clm-fates

### ISSUE THREAD

**Issue by glemieux** [glemieux/E3SM#18]
Subtasks:

- [x] Adapt base type for fates fire data types and method
- [x] Add elmfates_interface initialization code
- [x] refactor existing elm code to make sure cnfire and spitfire are separate
- [x] Incorporate `need_lightning_and_popdens`
- [x] Pass `gdp_lf_col`, `forc_lnfm`, and `forc_hdm` into fates fire types
- [x] Move need lightnting set function into fates fire base type
~- Add `surfacedataread` procedure~

**Comment by glemieux** [comment]
elmfates_interface initialization procedures:
- [x] `InterpFileInputs`
- [x] `Init2`
- [x] `InitAccBuffer`
- [x] `InitAccVars`
- [x] `UpdateAccVars`

**Comment by glemieux** [comment]
- `InterpFileInputs` is behind an `else` statement that is a result of a `use_cn` check.  

CLM-side notes:
- the `clm_fates%InitAccBuffer(bounds)` and `alm_fates%Init2` calls are behind a spitfire mode check, but the others are not.  Is this because the spitfire mode check is not available during the init, or it's just good practice to be on generally even if we aren't using spitfire? 
- there is a duplicate `use` call for the spitfire mode setting variables in `dynamics_driv`

**Comment by glemieux** [comment]
Check the calls to cnfire procedures to make sure that they won't be called with fates spitfire:
- [x] `FireInit`
- [x] `FireInterp`
- [x] `FireArea`
- [x] `FireFluxes`

**Comment by glemieux** [comment]
- `FireInit` and `FireInterp` are made available to the elmfates_interface as part of the method.  **This will need to be added somehow.**  Might just duplicate.
- `FireInit` is called within `EcosystemDynInit`, but if `use_fates` this will `return` out of the subroutine before being called.
- `FireArea` and `FireFluxes` is called by `CNEcosystemDynBetr` and `EcosystemDynNoLeaching2`.   Both procedures use `.not.use_fates` to skip the call.

**Comment by glemieux** [comment]
`FireInit` and `FireInterp` handling:
- the clm side is checking for `need_lightning_and_popdens` in the **base** fire data type to see whether or not it is needed.  This is a `deferred` procedure that is set by the upstream data types (cn/fates and data/nodata).  Bring this into FatesDataType
- it also calls `surfdataread` which I added during the refactor work on the [gdp refactor](https://github.com/ESCOMP/CTSM/pull/1536) to move stuff out of the CNVegStateType.  If I adapt this for elm-fates, but don't remove from CNVegStateType for now.
- `gdp_lf_col`, `forc_lnfm`, and `forc_hdm` are not allocated by the `FireInit` call on the elm-side.  They are allocated in the CNStateType, which where it was on the clm side before being refactored due to the polymorphic type update.

**Comment by glemieux** [comment]
- [ ] Add `surfacedataread` procedure
- [ ] Determine how to incorporate `need_lightning_and_popdens`

**Comment by glemieux** [comment]
`surfacedataread` handling:
- it looks like CNStateType that calls the `ncd_io` input functions are not behind a `use_cn` check (only the type's `InitHistory` is behind such a check), so should be generally available regardless of fire model.
- Note that although the `forc_hdm` and `forc_lnfm` data are initialized in FireMod.F90 with `FireInit` they are `atm2lnd_vars` type values!  I double checked the initialization route and confirmed that this routine is called prior to `atm2lnd_vars%forc_lnfm` setting.

**Comment by glemieux** [comment]
`CPL_BYPASS` handling:

I was confused on what the check for this compiler flag was doing so I asked Peter and Gautam for help and they pointed me to Dan.  Dan helped clarify (via slack) that these are two meant as two separate methods for providing access to the forcing data:

> The lightning and hdm data (forc_lnfm and forc_hdm) are read in and interpolated in lnd_import_export.F90 instead of using the init and interp functions in FireMod.  It uses the same stream file name.

Charlie provided additional context that this is used to avoid slowdowns.  Dan alludes to this and clarifies further that: 

> FATES should be able to run either way. However because FATES is more expensive than regular ELM, the benefit of using the bypass is smaller.  There are a few namelist variables that need to be set if you are using CPL_BYPASS, as well as setting the flag in cmake_macros/universal.cmake. You can look at OLMT (https://github.com/dmricciuto/OLMT/) and how it sets variables if cpl_bypass is True.

So the general flowchart for the forcing data is as follows:
- `#ifdef CPL_BYPASS` then initialize and interpolate `forc_lnfm` (and others) within the atm2lndType module.
- `#ifndef CPL_BYPASS` then initialize and interpolate *local versions of* `forc_lnfm` (and others) with `FireInit` and `FireInterp` subcalls (`lnfm_init`, etc)

The big-leaf FireMod code then handles this by using an additional check on `CPL_BYPASS` to `associate` the local variables with the `atm2lnd_vars` type if it is set.  The issue for fates is that while the `CPL_BYPASS=true` option can be handled pretty straight-forwaldly by accessing `atm2lnd_vars`, the non-bypass case the variables of interest are local to FireMod currently.  

So my initial thoughts on the broad options for handling this are:
- Refactor the FireMod.F90 code to add a `fire_base_type` to encapsulate the data to enable the fates fire module to use this (will this clash with the `associate` statement?
- Add spitfire mode checks to the `FireInit` and `FireInterp` sub-calls to make the data accessible to FATES fire module (problematic due to circular dependencies?)

**Comment by glemieux** [comment]
> So my initial thoughts on the broad options for handling this are:
> 
> * Refactor the FireMod.F90 code to add a `fire_base_type` to encapsulate the data to enable the fates fire module to use this (will this clash with the `associate` statement?
> * Add spitfire mode checks to the `FireInit` and `FireInterp` sub-calls to make the data accessible to FATES fire module (problematic due to circular dependencies?)

I'm realizing I should check with Peter Schwartz about how difficult refactoring FireMod would be wrt impact on the `!$acc ` gpu calls.  Maybe instead of refactoring the `forc_lnfm` code I simply add a super-procedure that copies the local private data into a the fates fire base type.

I think ultimately, it will be simpler to replicate `FireInit` and `FireInterp` routines in the `fates_fire_base_type` instead.

**Comment by glemieux** [comment]
> I'm realizing I should check with Peter Schwartz about how difficult refactoring FireMod would be wrt impact on the `!$acc ` gpu calls. Maybe instead of refactoring the `forc_lnfm` code I simply add a super-procedure that copies the local private data into a the fates fire base type.
> 
> I think ultimately, it will be simpler to replicate `FireInit` and `FireInterp` routines in the `fates_fire_base_type` instead.

I decided to take a mixed approach here by adapting the clm fire method and base type modules into elm alongside the existing fire module, so as to avoid refactoring the existing code.  The idea here is that while this duplicates code (specifically the procedures in the `FireInit` and `FireInterp` calls), it sets up the elm fire module to adopt this factory methodology in the future should it be desired, without having to refactor the fire module right now.

**Comment by glemieux** [comment]
`hdm_init` comparison

- elm: `gsmap=gsmap_lnd_gdc2glo`, clm: `gsmap=gsmap_global`

`lnfm_init` comparison:

- `tintalgo=lightng_tintalgo` is not used as an option in `shr_strdata_create` for the original FireMod

**Comment by glemieux** [comment]
I'm running into what I think is a circular dependency issue as the error message that I'm getting from the build doesn't make a lot of sense at first glance since I haven't updated CH4Mod:
```
/home/glemieux/Repos/E3SM/e3sm/components/elm/src/biogeochem/CH4Mod.F90:40:6:

   40 |   use ELMFatesInterfaceMod, only : hlm_fates_interface_type
      |      1
Fatal Error: Cannot open module file ‘elmfatesinterfacemod.mod’ for reading at (1): No such file or directory
compilation terminated.
```

My initial thoughts were that it had to do with how I was passing `cnstate_type` variables through the `Init2` subroutine calls, but that was a bust.

**Comment by glemieux** [comment]
I missed that the circular dependency error is listed directly in the bldlog: 
```
make[2]: Circular cmake/lnd/CMakeFiles/lnd.dir/__/__/elm/src/main/controlMod.F90.o <- cmake/lnd/CMakeFiles/lnd.dir/allocationmod.mod.stamp dependency dropped.
...
make[2]: Circular cmake/lnd/CMakeFiles/lnd.dir/__/__/elm/src/biogeochem/CH4Mod.F90.o <- cmake/lnd/CMakeFiles/lnd.dir/elmfatesinterfacemod.mod.stamp dependency dropped.
```

This means that a prerequisite of allocationMod depends on controlMod and a prerequisite of elmfatesinterfacemod depends on ch4mod.

**Comment by glemieux** [comment]
`use` of `elmfatesinterfacemod`

AllocationMod.F90:
- `AllocationInit` (dead code or standby?) <- `EcosystemDynInit` <- `initialize2`
- `Allocation2_ResolveNPLimit`

CH4Mod.F90
- `CH4`

elm_initializeMod.F90
- `initialize1`
- `initialize2`

elm_instMod.F90
- `elm_inst_biogeochem`

**Comment by glemieux** [comment]
Starting from `controlmod`, I determined that the issue appears to be due to the fact that the fates copy of the `hdm_init` and `lnfm_init`: 

![graph (3)](https://user-images.githubusercontent.com/7565064/211419064-8b6952fa-3503-41d9-b5f8-96867140ac66.png)

I'll need to see if there is some way to work around this.  The main issue is that `allocationMod` is using `elmfatesinterfaceMod` and `controlmod` is using `alloctionMod`, both of which are not present in clm.  The `controlMod` is using `AllocationMod` for public mode settings `suplnitro` and `suplphos` which have been located elsewhere in clm.

**Comment by glemieux** [comment]
Trying to break the dependency between `elmfatesinterfacemod` and `AllocationMod`, I think I discovered why it was being used.  The problem is that `AllocationMod` is being used by `controlMod`.  So even if I trying an setup `AllocationMod` to use `elm_instMod` to gain access to `alm_fates` instead of using `elmfatesinterfacemod` directly, I'll create an even tighter circular dependency (upper left subgraph):

![graph (4)](https://user-images.githubusercontent.com/7565064/211426159-2dbdacb3-4d02-4e3c-aa74-ba788546283b.png)

This is not to mention that there is still an existing circular dependency through `ch4Mod` by way of `ndepStreaMod`.  This would still be the case even with a refactor of the existing `FireMod` as `elmfatesinterfacemod` would still have to use `FireMod` which would complete the circular dependency.

Based on this, I think the simplest thing to do is to break the dependency between `AllocationMod` and `controlMod`.  That said, this is really an issue with the fates nutrients code as is currently designed and only doesn't show up in clm as it hasn't been ported there yet (the equivalent is `SoilBiogeochemCompetitionMod`).

**Comment by glemieux** [comment]
I just realized that I could more easily break the dependency through `ndepStreamsMod` by converting the `use controlMod, only : NLFilename` statement to an input argument for `ndep_init` where the filename is just a character input.  This has widespread precedent elsewhere in the code.

**Comment by glemieux** [comment]
This builds successfully now.  Make sure to update the design document with the findings above.

**Comment by glemieux** [comment]
> This builds successfully now. Make sure to update the design document with the findings above.

https://github.com/NGEET/fates-users-guide/pull/32

**Comment by glemieux** [comment]
I did not implement the `surfdataread` procedure as it is currently only necessary for spitfire modes greater than `3`.  It should be implemented when those modes are needed in the future.
